sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -53,6 +53,17 @@ case class UnaryMinus(
   override def toString: String = s"-$child"
 
   private lazy val numeric = TypeUtils.getNumeric(dataType, failOnError)
+  private lazy val unaryMinusFunc: Any => Any = dataType match {
```

Comment:
Why does this need to be a lazy val instead of just a def? I'm not sure why the line above is lazy either

Reply:
hmm, this would be called inside `eval` so a cached function is expected to avoid invoke by every row

Reply:
I mean, why declare it this way instead of a function? I don't see what difference it makes. There is nothing expensive about creating it, so why lazy?

Reply:
Do you mean rewrite it to ?
```scala
private def unaryMinusFunc: Any => Any = dataType match {
  ......
```

The reason I pull out and make it as lazy val is: the data type is known before do eval in an expression. Let's say if the data type is integer,
- then the function can be elimiated to `input  => numeric.negate(input)` during execution
- if you declare it as a function, then the function would be during execution:
  ```scala
  dataType match {
    ....
    case _ => input => numeric.negate(input)
  ```

The overhead if not about creating a function, but the data type matching elimination. So I think the lazy val is more efficient ?

Reply:
Oh I see now it depends on dataType. I understand what this does now.
How many other places would you have to change though to cover all functions? or are these the important ones?

Reply:
@srowen @LuciferYang  
This issue mostly happens in arithmetic like expression, and I also tried to collect other related expressions. Hope I'm not missing someone. We can still keep in mind for catching and reviewing the added expression or the new supported data type which may introduce this issue.

Reply:
What is the performance improvement like? I think we'd normally ask for a benchmark. It's not a big change but not trivial; if it's a good win maybe it's worthwhile doing for more. Any numbers?